File: ios-app/PracticeFlux/Core/Models/Appointment.swift
import Foundation

// MARK: - Appointment Model

struct Appointment: Codable, Identifiable, Equatable {
    let id: String
    let patientId: String
    let providerId: String
    let locationId: String?
    let appointmentType: AppointmentType
    let status: AppointmentStatus
    let startTime: Date
    let endTime: Date
    let notes: String?
    let reason: String?
    let isVirtual: Bool
    let virtualMeetingUrl: String?
    let provider: Provider?
    let location: Location?
    let createdAt: Date
    let updatedAt: Date

    var duration: TimeInterval {
        endTime.timeIntervalSince(startTime)
    }

    var durationMinutes: Int {
        Int(duration / 60)
    }

    var isUpcoming: Bool {
        startTime > Date()
    }

    var isPast: Bool {
        endTime < Date()
    }

    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        return formatter.string(from: startTime)
    }

    var formattedTime: String {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return "\(formatter.string(from: startTime)) - \(formatter.string(from: endTime))"
    }

    enum CodingKeys: String, CodingKey {
        case id, status, notes, reason, provider, location
        case patientId = "patient_id"
        case providerId = "provider_id"
        case locationId = "location_id"
        case appointmentType = "appointment_type"
        case startTime = "start_time"
        case endTime = "end_time"
        case isVirtual = "is_virtual"
        case virtualMeetingUrl = "virtual_meeting_url"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

enum AppointmentType: String, Codable, CaseIterable {
    case checkup = "CHECKUP"
    case followUp = "FOLLOW_UP"
    case consultation = "CONSULTATION"
    case procedure = "PROCEDURE"
    case labWork = "LAB_WORK"
    case imaging = "IMAGING"
    case vaccination = "VACCINATION"
    case emergency = "EMERGENCY"
    case telehealth = "TELEHEALTH"

    var displayName: String {
        switch self {
        case .checkup: return "Check-up"
        case .followUp: return "Follow-up"
        case .consultation: return "Consultation"
        case .procedure: return "Procedure"
        case .labWork: return "Lab Work"
        case .imaging: return "Imaging"
        case .vaccination: return "Vaccination"
        case .emergency: return "Emergency"
        case .telehealth: return "Telehealth"
        }
    }

    var icon: String {
        switch self {
        case .checkup: return "stethoscope"
        case .followUp: return "arrow.counterclockwise"
        case .consultation: return "person.2"
        case .procedure: return "cross.case"
        case .labWork: return "testtube.2"
        case .imaging: return "camera.metering.unknown"
        case .vaccination: return "syringe"
        case .emergency: return "cross.circle"
        case .telehealth: return "video"
        }
    }
}

enum AppointmentStatus: String, Codable, CaseIterable {
    case scheduled = "SCHEDULED"
    case confirmed = "CONFIRMED"
    case checkedIn = "CHECKED_IN"
    case inProgress = "IN_PROGRESS"
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"
    case noShow = "NO_SHOW"
    case rescheduled = "RESCHEDULED"

    var displayName: String {
        switch self {
        case .scheduled: return "Scheduled"
        case .confirmed: return "Confirmed"
        case .checkedIn: return "Checked In"
        case .inProgress: return "In Progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .noShow: return "No Show"
        case .rescheduled: return "Rescheduled"
        }
    }

    var color: String {
        switch self {
        case .scheduled: return "blue"
        case .confirmed: return "green"
        case .checkedIn: return "purple"
        case .inProgress: return "orange"
        case .completed: return "gray"
        case .cancelled: return "red"
        case .noShow: return "red"
        case .rescheduled: return "yellow"
        }
    }
}

// MARK: - Provider Model

struct Provider: Codable, Identifiable, Equatable {
    let id: String
    let userId: String
    let specialty: String
    let licenseNumber: String?
    let npi: String?
    let user: User?

    var displayName: String {
        if let user = user {
            return "Dr. \(user.fullName)"
        }
        return "Provider"
    }

    enum CodingKeys: String, CodingKey {
        case id, specialty, user
        case userId = "user_id"
        case licenseNumber = "license_number"
        case npi
    }
}

// MARK: - Location Model

struct Location: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let address: String
    let city: String
    let state: String
    let zipCode: String
    let phone: String?
    let isActive: Bool

    var fullAddress: String {
        "\(address), \(city), \(state) \(zipCode)"
    }

    enum CodingKeys: String, CodingKey {
        case id, name, address, city, state, phone
        case zipCode = "zip_code"
        case isActive = "is_active"
    }
}

// MARK: - Appointment Requests

struct CreateAppointmentRequest: Codable {
    let providerId: String
    let locationId: String?
    let appointmentType: AppointmentType
    let startTime: Date
    let endTime: Date
    let reason: String?
    let isVirtual: Bool

    enum CodingKeys: String, CodingKey {
        case reason
        case providerId = "provider_id"
        case locationId = "location_id"
        case appointmentType = "appointment_type"
        case startTime = "start_time"
        case endTime = "end_time"
        case isVirtual = "is_virtual"
    }
}

struct RescheduleAppointmentRequest: Codable {
    let startTime: Date
    let endTime: Date
    let reason: String?

    enum CodingKeys: String, CodingKey {
        case reason
        case startTime = "start_time"
        case endTime = "end_time"
    }
}

File: flutter-app/lib/core/models/appointment.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../theme/app_colors.dart';

enum AppointmentType {
  checkup,
  followUp,
  consultation,
  procedure,
  labWork,
  imaging,
  vaccination,
  emergency,
  telehealth,
}

enum AppointmentStatus {
  scheduled,
  confirmed,
  checkedIn,
  inProgress,
  completed,
  cancelled,
  noShow,
  rescheduled,
}

extension AppointmentTypeExtension on AppointmentType {
  String get displayName {
    switch (this) {
      case AppointmentType.checkup:
        return 'Check-up';
      case AppointmentType.followUp:
        return 'Follow-up';
      case AppointmentType.consultation:
        return 'Consultation';
      case AppointmentType.procedure:
        return 'Procedure';
      case AppointmentType.labWork:
        return 'Lab Work';
      case AppointmentType.imaging:
        return 'Imaging';
      case AppointmentType.vaccination:
        return 'Vaccination';
      case AppointmentType.emergency:
        return 'Emergency';
      case AppointmentType.telehealth:
        return 'Telehealth';
    }
  }

  IconData get icon {
    switch (this) {
      case AppointmentType.checkup:
        return Icons.medical_services_outlined;
      case AppointmentType.followUp:
        return Icons.replay;
      case AppointmentType.consultation:
        return Icons.people_outline;
      case AppointmentType.procedure:
        return Icons.healing_outlined;
      case AppointmentType.labWork:
        return Icons.science_outlined;
      case AppointmentType.imaging:
        return Icons.camera_alt_outlined;
      case AppointmentType.vaccination:
        return Icons.vaccines_outlined;
      case AppointmentType.emergency:
        return Icons.emergency_outlined;
      case AppointmentType.telehealth:
        return Icons.videocam_outlined;
    }
  }
}

extension AppointmentStatusExtension on AppointmentStatus {
  String get displayName {
    switch (this) {
      case AppointmentStatus.scheduled:
        return 'Scheduled';
      case AppointmentStatus.confirmed:
        return 'Confirmed';
      case AppointmentStatus.checkedIn:
        return 'Checked In';
      case AppointmentStatus.inProgress:
        return 'In Progress';
      case AppointmentStatus.completed:
        return 'Completed';
      case AppointmentStatus.cancelled:
        return 'Cancelled';
      case AppointmentStatus.noShow:
        return 'No Show';
      case AppointmentStatus.rescheduled:
        return 'Rescheduled';
    }
  }

  Color get color {
    switch (this) {
      case AppointmentStatus.scheduled:
        return AppColors.statusScheduled;
      case AppointmentStatus.confirmed:
        return AppColors.statusConfirmed;
      case AppointmentStatus.checkedIn:
        return AppColors.statusCheckedIn;
      case AppointmentStatus.inProgress:
        return AppColors.statusInProgress;
      case AppointmentStatus.completed:
        return AppColors.statusCompleted;
      case AppointmentStatus.cancelled:
      case AppointmentStatus.noShow:
        return AppColors.statusCancelled;
      case AppointmentStatus.rescheduled:
        return AppColors.warning;
    }
  }
}

class Provider extends Equatable {
  final String id;
  final String userId;
  final String specialty;
  final String? licenseNumber;
  final String? npi;
  final String? firstName;
  final String? lastName;

  const Provider({
    required this.id,
    required this.userId,
    required this.specialty,
    this.licenseNumber,
    this.npi,
    this.firstName,
    this.lastName,
  });

  String get displayName {
    if (firstName != null && lastName != null) {
      return 'Dr. $firstName $lastName';
    }
    return 'Provider';
  }

  factory Provider.fromJson(Map<String, dynamic> json) {
    final user = json['user'] as Map<String, dynamic>?;
    return Provider(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      specialty: json['specialty'] as String,
      licenseNumber: json['license_number'] as String?,
      npi: json['npi'] as String?,
      firstName: user?['first_name'] as String?,
      lastName: user?['last_name'] as String?,
    );
  }

  @override
  List<Object?> get props => [id, userId, specialty, licenseNumber, npi];
}

class Location extends Equatable {
  final String id;
  final String name;
  final String address;
  final String city;
  final String state;
  final String zipCode;
  final String? phone;
  final bool isActive;

  const Location({
    required this.id,
    required this.name,
    required this.address,
    required this.city,
    required this.state,
    required this.zipCode,
    this.phone,
    required this.isActive,
  });

  String get fullAddress => '$address, $city, $state $zipCode';

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      id: json['id'] as String,
      name: json['name'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      state: json['state'] as String,
      zipCode: json['zip_code'] as String,
      phone: json['phone'] as String?,
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  @override
  List<Object?> get props => [id, name, address, city, state, zipCode];
}

class Appointment extends Equatable {
  final String id;
  final String patientId;
  final String providerId;
  final String? locationId;
  final AppointmentType appointmentType;
  final AppointmentStatus status;
  final DateTime startTime;
  final DateTime endTime;
  final String? notes;
  final String? reason;
  final bool isVirtual;
  final String? virtualMeetingUrl;
  final Provider? provider;
  final Location? location;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Appointment({
    required this.id,
    required this.patientId,
    required this.providerId,
    this.locationId,
    required this.appointmentType,
    required this.status,
    required this.startTime,
    required this.endTime,
    this.notes,
    this.reason,
    required this.isVirtual,
    this.virtualMeetingUrl,
    this.provider,
    this.location,
    required this.createdAt,
    required this.updatedAt,
  });

  Duration get duration => endTime.difference(startTime);
  int get durationMinutes => duration.inMinutes;
  bool get isUpcoming => startTime.isAfter(DateTime.now());
  bool get isPast => endTime.isBefore(DateTime.now());

  String get formattedDate => DateFormat('MMM d, yyyy').format(startTime);
  String get formattedTime =>
      '${DateFormat('h:mm a').format(startTime)} - ${DateFormat('h:mm a').format(endTime)}';

  factory Appointment.fromJson(Map<String, dynamic> json) {
    return Appointment(
      id: json['id'] as String,
      patientId: json['patient_id'] as String,
      providerId: json['provider_id'] as String,
      locationId: json['location_id'] as String?,
      appointmentType: AppointmentType.values.firstWhere(
        (e) =>
            e.name.toUpperCase() ==
            (json['appointment_type'] as String).replaceAll('_', '').toUpperCase(),
        orElse: () => AppointmentType.checkup,
      ),
      status: AppointmentStatus.values.firstWhere(
        (e) =>
            e.name.toUpperCase() ==
            (json['status'] as String).replaceAll('_', '').toUpperCase(),
        orElse: () => AppointmentStatus.scheduled,
      ),
      startTime: DateTime.parse(json['start_time'] as String),
      endTime: DateTime.parse(json['end_time'] as String),
      notes: json['notes'] as String?,
      reason: json['reason'] as String?,
      isVirtual: json['is_virtual'] as bool? ?? false,
      virtualMeetingUrl: json['virtual_meeting_url'] as String?,
      provider: json['provider'] != null
          ? Provider.fromJson(json['provider'] as Map<String, dynamic>)
          : null,
      location: json['location'] != null
          ? Location.fromJson(json['location'] as Map<String, dynamic>)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patient_id': patientId,
      'provider_id': providerId,
      'location_id': locationId,
      'appointment_type': appointmentType.name.toUpperCase(),
      'status': status.name.toUpperCase(),
      'start_time': startTime.toIso8601String(),
      'end_time': endTime.toIso8601String(),
      'notes': notes,
      'reason': reason,
      'is_virtual': isVirtual,
      'virtual_meeting_url': virtualMeetingUrl,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  @override
  List<Object?> get props => [
        id,
        patientId,
        providerId,
        locationId,
        appointmentType,
        status,
        startTime,
        endTime,
        isVirtual,
      ];
}
